File: code/data_structures/indexed_priority_queue/MinIndexDHeap.cpp


#include <vector>
#include <stdexcept>
#include <algorithm> // for std::min and std::max

template<typename T>
class MinIndexedDHeap {
private:
    std::vector<T> values;
    std::vector<int> pm, im, child, parent;
    int sz, N, D;

public:
    MinIndexedDHeap(int degree, int maxSize) : D(std::max(2, degree)), N(std::max(D + 1, maxSize)), values(N), pm(N), im(N), child(N), parent(N), sz(0) {
        for (int i = 0; i < N; ++i) {
            parent[i] = (i - 1) / D;
            child[i] = i * D + 1;
            pm[i] = im[i] = -1;
        }
    }

    int size() {
        return sz;
    }

    bool isEmpty() {
        return sz == 0;
    }

    bool contains(int ki) {
        keyInBoundsOrThrow(ki);
        return pm[ki] != -1;
    }

    int peekMinKeyIndex() {
        isNotEmptyOrThrow();
        return im[0];
    }

    int pollMinKeyIndex() {
        int minki = peekMinKeyIndex();
        deleteKey(minki);
        return minki;
    }

    T peekMinValue() {
        isNotEmptyOrThrow();
        return values[im[0]];
    }

    T pollMinValue() {
        T minValue = peekMinValue();
        deleteKey(peekMinKeyIndex());
        return minValue;
    }

    void insert(int ki, T value) {
        if (contains(ki)) throw std::invalid_argument("index already exists; received: " + ki);
        valueNotNullOrThrow(value);
        pm[ki] = sz;
        im[sz] = ki;
        values[ki] = value;
        swim(sz++);
    }

    T valueOf(int ki) {
        keyExistsOrThrow(ki);
        return values[ki];
    }

    T deleteKey(int ki) {
        keyExistsOrThrow(ki);
        int i = pm[ki];
        swap(i, --sz);
        sink(i);
        swim(i);
        T value = values[ki];
        values[ki] = T();
        pm[ki] = -1;
        im[sz] = -1;
        return value;
    }

    T update(int ki, T value) {
        keyExistsAndValueNotNullOrThrow(ki, value);
        int i = pm[ki];
        T oldValue = values[ki];
        values[ki] = value;
        sink(i);
        swim(i);
        return oldValue;
    }

    void decrease(int ki, T value) {
        keyExistsAndValueNotNullOrThrow(ki, value);
        if (less(value, values[ki])) {
            values[ki] = value;
            swim(pm[ki]);
        }
    }

    void increase(int ki, T value) {
        keyExistsAndValueNotNullOrThrow(ki, value);
        if (less(values[ki], value)) {
            values[ki] = value;
            sink(pm[ki]);
        }
    }

    bool isMinHeap() {
        return isMinHeap(0);
    }

private:
    void sink(int i) {
        int j = minChild(i);
        while (j != -1) {
            swap(i, j);
            i = j;
            j = minChild(i);
        }
    }

    void swim(int i) {
        while (less(i, parent[i])) {
            swap(i, parent[i]);
            i = parent[i];
        }
    }

    int minChild(int i) {
        int index = -1, from = child[i], to = std::min(sz, from + D);
        for (int j = from; j < to; ++j) {
            if (less(j, i)) {
                index = i = j;
            }
        }
        return index;
    }

    void swap(int i, int j) {
        pm[im[j]] = i;
        pm[im[i]] = j;
        std::swap(im[i], im[j]);
    }

    bool less(int i, int j) {
        return values[im[i]] < values[im[j]];
    }

    bool less(T obj1, T obj2) {
        return obj1 < obj2;
    }

    void isNotEmptyOrThrow() {
        if (isEmpty()) throw std::out_of_range("Priority queue underflow");
    }

    void keyExistsAndValueNotNullOrThrow(int ki, T value) {
        keyExistsOrThrow(ki);
        valueNotNullOrThrow(value);
    }

    void keyExistsOrThrow(int ki) {
        if (!contains(ki)) throw std::out_of_range("Index does not exist; received: " + ki);
    }

    void valueNotNullOrThrow(T value) {
        if (value == NULL) throw std::invalid_argument("value cannot be null");
    }

    void keyInBoundsOrThrow(int ki) {
        if (ki < 0 || ki >= N)
            throw std::invalid_argument("Key index out of bounds; received: " + ki);
    }

    bool isMinHeap(int i) {
        int from = child[i], to = std::min(sz, from + D);
        for (int j = from; j < to; ++j) {
            if (!less(i, j)) return false;
            if (!isMinHeap(j)) return false;
        }
        return true;
    }
};

File: code/data_structures/indexed_priority_queue/MinIndexDHeap.py
from math import min
from typing import Any, List, Union

from collections.abc import NoSuchElementException


class MinIndexedDHeap:
    def __init__(self, degree: int, maxSize: int):
        if maxSize <= 0:
            raise ValueError("maxSize <= 0")

        self.D = max(2, degree)
        self.N = max(self.D + 1, maxSize)

        # Create unfilled inverse and position maps
        # and the values array. We also create arrays
        # where we will store how to reach of the child 
        # and parent nodes for each node.
        self.im = [-1] * self.N
        self.pm = [-1] * self.N
        self.child = [0] * self.N
        self.parent = [0] * self.N
        self.values = [None] * self.N

        for i in range(self.N):
            self.parent[i] = (i - 1) // self.D
            self.child[i] = i * self.D + 1

        # Size of the heap, the number of nodes it contains
        self.sz = 0

    def size(self) -> int:
        return self.sz

    def isEmpty(self) -> bool:
        return self.sz == 0

    # We will check if the key index of the wanted
    # element is not -1 
    def contains(self, ki: int) -> bool:
        self.keyInBoundsOrThrow(ki)
        return self.pm[ki] != -1

    def peekMinKeyIndex(self) -> int:
        self.isNotEmptyOrThrow()
        return self.im[0]

    def pollMinKeyIndex(self) -> int:
        minki = self.peekMinKeyIndex()
        self.delete(minki)
        return minki

    def peekMinValue(self) -> Any:
        self.isNotEmptyOrThrow()
        return self.values[self.im[0]]

    def pollMinValue(self) -> Any:
        minValue = self.peekMinValue()
        self.delete(self.peekMinKeyIndex())
        return minValue

    # We first insert the new node at the bottom right position,
    # which is conveniently given by the size of the heap. Insert
    # it there and then bubble up the node until we preserve the
    # heap invariance. When we swap nodes we have to take into
    # account that we have to change the values in the position
    # and inverse maps. 
    def insert(self, ki: int, value: Any) -> None:
        if self.contains(ki):
            raise ValueError("index already exists; received: " + str(ki))
        self.valueNotNullOrThrow(value)
        self.pm[ki] = self.sz
        self.im[self.sz] = ki
        self.values[ki] = value
        self.swim(self.sz)
        self.sz += 1

    def valueOf(self, ki: int) -> Any:
        self.keyExistsOrThrow(ki)
        return self.values[ki]

    # For deletion, we first swap our node with
    # the lower right node of the heap and then
    # delete it. Then we have to preserve the heap
    # invariant, so we have to bubble up or bubble down
    # the node which previously was on the lower right.
    # We use both, sink and swim. If it can swim it will,
    # otherwise it will sink, and otherwise it will mean
    # that when doing the change we already satisfied
    # the heap invariant
    def delete(self, ki: int) -> Any:
        self.keyExistsOrThrow(ki)
        i = self.pm[ki]
        self.swap(i, self.sz - 1)
        self.sz -= 1
        self.sink(i)
        self.swim(i)
        value = self.values[ki]
        self.values[ki] = None
        self.pm[ki] = -1
        self.im[self.sz] = -1
        return value
    
    # We can update the valuesof our elements, so, when doing
    # this we have to adapt the values, pm and im arrays
    def update(self, ki: int, value: Any) -> Any:
        self.keyExistsAndValueNotNullOrThrow(ki, value)
        i = self.pm[ki]
        oldValue = self.values[ki]
        self.values[ki] = value
        self.sink(i)
        self.swim(i)
        return oldValue

    def decrease(self, ki: int, value: Any) -> None:
        self.keyExistsAndValueNotNullOrThrow(ki, value)
        if self.less(value, self.values[ki]):
            self.values[ki] = value
            self.swim(self.pm[ki])

    def increase(self, ki: int, value: Any) -> None:
        self.keyExistsAndValueNotNullOrThrow(ki, value)
        if self.less(self.values[ki], value):
            self.values[ki] = value
            self.sink(self.pm[ki])

    def sink(self, i: int) -> None:
        while True:
            j = self.minChild(i)
            if j == -1:
                break
            self.swap(i, j)
            i = j

    def swim(self, i: int) -> None:
        while i > 0 and self.less(i, self.parent[i]):
            self.swap(i, self.parent[i])
            i = self.parent[i]

    def minChild(self, i: int) -> int:
        index = -1
        from_idx = self.child[i]
        to_idx = min(self.sz, from_idx + self.D)
        for j in range(from_idx, to_idx):
            if self.less(j, i):
                index = i = j
        return index

    def swap(self, i: int, j: int) -> None:
        self.pm[self.im[j]] = i
        self.pm[self.im[i]] = j
        self.im[i], self.im[j] = self.im[j], self.im[i]

    def less(self, i: int, j: int) -> bool:
        return self.values[self.im[i]] < self.values[self.im[j]]

    def isMinHeap(self) -> bool:
        return self.isMinHeapRecursive(0)

    def isMinHeapRecursive(self, i: int) -> bool:
        from_idx = self.child[i]
        to_idx = min(self.sz, from_idx + self.D)
        for j in range(from_idx, to_idx):
            if not self.less(i, j):
                return False
            if not self.isMinHeapRecursive(j):
                return False
        return True

    def isNotEmptyOrThrow(self) -> None:
        if self.isEmpty():
            raise NoSuchElementException("Priority queue underflow")

    def keyExistsAndValueNotNullOrThrow(self, ki: int, value: Any) -> None:
        self.keyExistsOrThrow(ki)
        self.valueNotNullOrThrow(value)

    def keyExistsOrThrow(self, ki: int) -> None:
        if not self.contains(ki):
            raise NoSuchElementException("Index does not exist; received: " + str(ki))

    def valueNotNullOrThrow(self, value: Any) -> None:
        if value is None:
            raise ValueError("value cannot be null")

    def keyInBoundsOrThrow(self, ki: int) -> None:
        if ki < 0 or ki >= self.N:
            raise ValueError("Key index out of bounds; received: " + str(ki))
